pkg/tfbridge/walk.go

```diff
@@ -168,3 +265,347 @@ func LookupSchemaInfoPath(schemaPath SchemaPath, schemaInfo *SchemaInfo) *Schema
 		return nil
 	}
 }
+
+// Drill down a path from a SchemaInfo object and find a matching SchemaInfo. If no
+// SchemaInfo exists, it will be created.
+func getSchemaInfoPath(schemaPath SchemaPath, schemaInfo *SchemaInfo) *SchemaInfo {
+	if len(schemaPath) == 0 {
+		return schemaInfo
+	}
+
+	switch first := schemaPath[0].(type) {
+	case walk.ElementStep:
+		return getSchemaInfoPath(schemaPath[1:], ensureField(&schemaInfo.Elem))
+	case walk.GetAttrStep:
+		return getSchemaInfoPath(schemaPath[1:], ensureMapKey(&schemaInfo.Fields, first.Name))
+	default:
+		contract.Failf("Unable to walk SchemaPath segment of type %T", schemaPath[0])
+		return nil
+	}
+}
+
+type VisitInfo struct {
+	Root VisitRoot
+
+	shimSchema    shim.Schema
+	getSchemaInfo func() *SchemaInfo
+
+	getShimSchemaMap func() shim.SchemaMap
+	getSchemaInfoMap func() map[string]*SchemaInfo
+}
+
+type VisitResult struct {
+	// If the visitor has done something.
+	//
+	// Visits will only be replayed during runtime startup if `HasEffect: true`.
+	HasEffect bool
+}
+
+func (v *VisitInfo) ShimSchema() shim.Schema                        { return v.shimSchema }
+func (v *VisitInfo) SchemaInfo() *SchemaInfo                        { return v.getSchemaInfo() }
+func (v *VisitInfo) EnclosingSchemaInfoMap() map[string]*SchemaInfo { return v.getSchemaInfoMap() }
+func (v *VisitInfo) EnclosingSchemaShimMap() shim.SchemaMap         { return v.getShimSchemaMap() }
+
+type VisitRoot interface {
+	PulumiToken() tokens.Type
+
+	isVisitRoot()
+}
+
+// This ensures that we can peform casts from VisitRoot to T for each valid root. Changing
+// this from T to *T is a breaking change.
+var (
+	_ VisitRoot = VisitResourceRoot{}
+	_ VisitRoot = VisitDataSourceRoot{}
+	_ VisitRoot = VisitConfigRoot{}
+)
+
+type VisitResourceRoot struct {
+	token tokens.Type
+
+	Info *ResourceInfo
+
+	TfToken string
+}
+
+func (r VisitResourceRoot) PulumiToken() tokens.Type { return r.token }
+func (r VisitResourceRoot) isVisitRoot()             {}
+
+type VisitDataSourceRoot struct {
+	token tokens.Type
+
+	Info *DataSourceInfo
+
+	TfToken string
+}
+
+func (r VisitDataSourceRoot) PulumiToken() tokens.Type { return r.token }
+func (r VisitDataSourceRoot) isVisitRoot()             {}
+
+type VisitConfigRoot struct{ token tokens.Type }
+
+func (r VisitConfigRoot) PulumiToken() tokens.Type { return r.token }
+func (r VisitConfigRoot) isVisitRoot()             {}
+
+type Visitor = func(VisitInfo, SchemaPath) (VisitResult, error)
```

Comment:
Style golf, perhaps simplify to `func(VisitContext) error`, VisitContext can make other things handy like methods to retrieve "current" schema or info.